### ISSUE TITLE
MSVC 2019 cl.exe builds are still not invariant

### ISSUE THREAD

**Issue by solidpixel** [ARM-software/astc-encoder#227]
With the latest changes to use invariant accumulators we've solved the output invariance issues we were seeing with clang++ and g++. This includes being self-invariant across SIMD instruction sets, invariant between GCC and LLVM, and invariant across operating systems. I've yet to thoroughly test invariance across x64 and Arm, but on a cursory skim it looks OK there too ... 

However, we're still seeing problems with Visual Studio builds using the `cl.exe` compiler. It is neither invariant with the other compilers, or invariant with itself across instruction sets. Given the goal of being invariant this needs investigating.

**Comment by solidpixel** [comment]
Testing with the LLVM toolset for MSV 2019, and it is a lot better but still not totally invariant with Clang for Linux.

**Comment by solidpixel** [comment]
The major differences are caused by projects defaulting to `/fp:precise` instead of `/fp:strict`. Setting compilers to strict mode at least makes `cl.exe` and `clangcl.exe` agree (but still disagree with GCC/Clang on Linux).

**Comment by solidpixel** [comment]
With this change only the following image and block sizes are invariant with Linux builds:

```
4x4 hdr-rgb-00.hdr (+0.002 dB)
8x8 hdr-rgb-00.hdr (-0.001 dB)
12x12 hdr-rgb-00.hdr (+0.004 dB)
```

**Comment by solidpixel** [comment]
Error fails to reproduce in a minified image for the impacted blocks, so suspect it's a precision issue in the summed-area variance tables.

**Comment by solidpixel** [comment]
Output of this block is different, so looks like we have different `std::powf()` outputs across compiler backends (which explains why `clang-cl.exe` is impacted, as IIUC it uses the same C library as `cl.exe`).

```
    d.set_lane<0>(powf(astc::max(d.lane<0>(), 1e-6f), rgb_power));
    d.set_lane<1>(powf(astc::max(d.lane<1>(), 1e-6f), rgb_power));
    d.set_lane<2>(powf(astc::max(d.lane<2>(), 1e-6f), rgb_power));
    d.set_lane<3>(powf(astc::max(d.lane<3>(), 1e-6f), alpha_power));
```

**Comment by solidpixel** [comment]
Confirmed that that seems to be the only problem, hacking in a fast (bad) approximation of `pow()` and using that everywhere resolves the remaining invariance issues. The error bars on the `pow()` implementation are large (max of ~25%) and we get relatively large IQ regressions, so its not a viable replacement, but it confirms what we need to replace ...

**Comment by solidpixel** [comment]
Interesting polynomial implementation here, which is MIT licensed: https://jrfonseca.blogspot.com/2008/09/fast-sse2-pow-tables-or-polynomials.html

**Comment by solidpixel** [comment]
Prototype of vector implementations for exp2/log2/pow added in PR #228. This results in invariance on my x86-64 device, but causes some regressions on the Apple M1, so I'm investigating that before merging.

**Comment by solidpixel** [comment]
Apple M1 native binaries for "none" are invariant with x86-64, but binaries for "neon" are not.

**Comment by solidpixel** [comment]
Fixed in bc18e8b
